R/get_source_expressions.R

```diff
@@ -18,32 +18,36 @@
-#'     \item{`column` (`integer`) the column in `filename` where this expression begins}
+#'     \item{`filename` (`character`) the name of the file.}
+#'     \item{`line` (`integer`) the line in the file where this expression begins.}
+#'     \item{`column` (`integer`) the column in the file where this expression begins.}
 #'     \item{`lines` (named `character`) vector of all lines spanned by this
-#'           expression, named with the line number corresponding to `filename`}
-#'     \item{`parsed_content` (`data.frame`) as given by [utils::getParseData()] for this expression}
-#'     \item{`xml_parsed_content` (`xml_document`) the XML parse tree of this
-#'          expression as given by [xmlparsedata::xml_parse_data()]}
-#'     \item{`content` (`character`) the same as `lines` as a single string (not split across lines)}
+#'           expression, named with the corresponding line numbers.}
+#'     \item{`parsed_content` (`data.frame`) as given by [utils::getParseData()] for this expression.}
+#'     \item{`xml_parsed_content` (`xml_document`) the XML parse tree of this expression as given by
+#'           [xmlparsedata::xml_parse_data()].}
+#'     \item{`content` (`character`) the same as `lines` as a single string (not split across lines).}
+#'     \item{`xml_find_function_calls(function_names)` (`function`) a function that returns all `SYMBOL_FUNCTION_CALL`
+#'           XML nodes from `xml_parsed_content` with specified function names.}
 #'   }
 #'
 #'   The final element of `expressions` is a list corresponding to the full file
-#'   consisting of 6 elements:
+#'   consisting of 7 elements:
 #'   \itemize{
-#'     \item{`filename` (`character`)}
-#'     \item{`file_lines` (`character`) the [readLines()] output for this file}
+#'     \item{`filename` (`character`) the name of this file.}
+#'     \item{`file_lines` (`character`) the [readLines()] output for this file.}
 #'     \item{`content` (`character`) for .R files, the same as `file_lines`;
-#'           for .Rmd or .qmd scripts, this is the extracted R source code (as text)}
+#'           for .Rmd or .qmd scripts, this is the extracted R source code (as text).}
 #'     \item{`full_parsed_content` (`data.frame`) as given by
-#'           [utils::getParseData()] for the full content}
+#'           [utils::getParseData()] for the full content.}
 #'     \item{`full_xml_parsed_content` (`xml_document`) the XML parse tree of all
-#'           expressions as given by [xmlparsedata::xml_parse_data()]}
+#'           expressions as given by [xmlparsedata::xml_parse_data()].}
 #'     \item{`terminal_newline` (`logical`) records whether `filename` has a terminal
-#'           newline (as determined by [readLines()] producing a corresponding warning)}
+#'           newline (as determined by [readLines()] producing a corresponding warning).}
+#'     \item{`xml_find_function_calls(function_names)` (`function`) a function that returns all `SYMBOL_FUNCTION_CALL`
```

Comment:
Not sure I like the approach of making this a factory, WDYT about just making this an entry with the named `xml_find_all()` output?

Another suggestion is making this a 3rd "linter level" (file>expression>call). Since these are pointers to nodes in the XML tree the overhead should not be crazy. OTOH copying the subsetted `getParseData()` data.frame will blow up the object size, but we could also pursue a breaking change whereby instead of fully subsetting the file-wide data.frame, the nested levels just consist of row indices on the full data.frame. So e.g. instead of a data.frame at expression level, we have an integer vector like `idx=c(1, 8, 12, 15)`, and `FullDF[idx,]` gives the subset we used to store earlier. Storing a bunch of integer vectors should be more efficient (and we hardly use this object anymore anyway). This is not a fully-fleshed-out idea but something kicking around in my head.

Would also like to start getting a sense of performance comparison. Normal (any random R file) and extreme (tools/R/QC.R) cases.

Reply:
The named call list has the disadvantage of exposing the implementation detail.
If we ever find a faster way to do it, or a fast way to add additional conditions (as often is done), we can't extend this feature.

re Performance: Sure, atm it's not integrated in any linter so will be slightly slower. Trying to suss out the API before creating a huge diff.

Reply:
NB a back-compatible change for smaller data structures could be using `rsample::resample()`. Not sure we should do that, tough.

Reply:
Happy with the current implementation, resolving.